lib/charms/alertmanager_k8s/v0/alertmanager_remote_configuration.py

```diff
@@ -0,0 +1,454 @@
+    ```
+
+    The `RemoteConfigurationProvider` provides handling of the most relevant charm
+    lifecycle events. On each of the defined Juju events, Alertmanager configuration and templates
+    from a specified file will be pushed to the relation data bag.
+    Inside the relation data bag, Alertmanager configuration will be stored under
+    `alertmanager_configuration` key, while the templates under the `alertmanager_templates` key.
+    Separation of the main configuration and the templates is dictated by the assumption that
+    the default provider of the `alertmanager_remote_configuration` relation will be
+    `alertmanager-k8s` charm, which requires such separation.
+    """
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        alertmanager_config: Optional[dict] = None,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ):
+        """API that manages a provided `remote-configuration` relation.
+
+        Args:
+            charm: The charm object that instantiated this class.
+            alertmanager_config: Alertmanager configuration dictionary.
+            relation_name: Name of the relation with the `alertmanager_remote_configuration`
+                interface as defined in metadata.yaml. Defaults to `remote-configuration`.
+        """
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self.alertmanager_config = alertmanager_config
+        self._relation_name = relation_name
+
+        on_relation = self._charm.on[self._relation_name]
+
+        self.framework.observe(on_relation.relation_joined, self._on_relation_joined)
+
+    @classmethod
+    def with_config_file(
+        cls,
+        charm: CharmBase,
+        config_file: Path,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ):
+        """The RemoteConfigurationProvider object factory.
+
+        This factory provides an alternative way of instantiating the RemoteConfigurationProvider.
+        While the default constructor requires passing a config dict, the factory allows using
+        a configuration file path.
+
+        Args:
+            charm: The charm object that instantiated this class.
+            config_file: Path to the Alertmanager configuration file.
+            relation_name: Name of the relation with the `alertmanager_remote_configuration`
+                interface as defined in metadata.yaml. Defaults to `remote-configuration`.
+
+        Returns:
+            RemoteConfigurationProvider object
+        """
+        return cls(charm, cls.load_config_file(config_file), relation_name)
+
+    def _on_relation_joined(self, _) -> None:
+        """Event handler for RelationJoinedEvent.
+
+        Takes care of pushing Alertmanager configuration to the relation data bag.
+        """
+        if not self._charm.unit.is_leader():
+            return
+        if self.alertmanager_config:
+            self.update_relation_data_bag(self.alertmanager_config)
```

Comment:
- Seems like the `if self.alertmanager_config:` can be moved inside `update_relation_data_bag`.
- Shouldn't `update_relation_data_bag` also clear reldata when `alertmanager_config` becomes falsey?

Reply:
I've completely removed this if. Validation will cover that anyway.
As for clearing the data bag - from the functional pov we don't have to do that. Also, the benefit of not doing that would be that we would have a valid configuration in the data bag at all times. If new Alertmanager relation joins, it can immediately fetch the configuration. WDYT?

Reply:
Need to think about this from a "juju model" POV. Rel-data, config options represent the model. Is it semantically correct to have stale data in this case? Not sure.

Wouldn't a new join generate the relation cascade?

Reply:
I agree with Leon here. If the relation is gone, the data should go as well as it otherwise gives a false representation of the juju model.